File: 1025-minimum-cost-for-tickets/minimum-cost-for-tickets.js
/**
 * @param {number[]} days
 * @param {number[]} costs
 * @return {number}
 */
var mincostTickets = function (days, costs) {
    const n = days.length;
    const last = days[n - 1];
    let dp = new Array(last + 1).fill(0);
    const setdays = new Set(days);

    for (let i = 1; i < last + 1; i++) {
        if (setdays.has(i)) {
            dp[i] = Math.min(
                costs[0] + dp[i - 1],
                costs[1] + dp[Math.max(i - 7, 0)],
                costs[2] + dp[Math.max(i - 30, 0)]);
        }
        else {
            dp[i] = dp[i - 1];
        }
    }

    return dp[last];
};

File: 1025-minimum-cost-for-tickets/minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        mx = max(days)
        dp = [0] * (mx + 1)
        days = set(days)

        for i in range(1, mx + 1):
            if i in days:
                dp[i] = min(costs[2] + dp[max(i - 30, 0)], costs[1] +  dp[max(i - 7, 0)], costs[0] + dp[i - 1])
            else:
                dp[i] = dp[i - 1]
            
        return dp[-1]